### PR TITLE
Color demo with ember collection and virtual each

### PR DESCRIPTION
**Pull request by flexyford**
This PR introduces the `virtual-each` (React-style-each) into `ember-impagination`. Therefore, the `impagination-collection` component interface has all functionality for `ember-collection` and `virtual-each`.

The `length` getter was tricky and for some unknown reason has to be explicitly defined outside the `create` of the `collection` Object

### CODE REVIEW

File: tests/dummy/app/templates/components/impagination-list.hbs [hunk 1]
```diff
@@ -0,0 +1 @@
+{{yield records state pages positionIndex}}
```
Comment (lydiaguarino) [inline tests/dummy/app/templates/components/impagination-list.hbs]: Should be removed.

Comment (flexyford) [inline tests/dummy/app/templates/components/impagination-list.hbs]: done

**Comment by cowboyd** [comment]
IMHO, `impagination-collection` is better named `imagination-dataset`. The reason is because `list` and `collection`, for better or for worse, already have visual connotations in the Ember world. There's `ember-collection`, and smoke-and-mirrors has `vertical-collection`, `horizontal-collection`, and `grid-collection`. A name like `-dataset` re-enforces the idea that this is only a data provider and is not visual in any way.

**Comment by cowboyd** [comment]
Looks like the build is failing on this branch.

File: addon/components/impagination-collection.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ import Dataset from 'ember-impagination/dataset';
 export default Ember.Component.extend({
   layout: layout,
   'initial-read-offset': 0,
-  'load-horizon': 1,
+  'load-horizon': 2,
```
Comment (cowboyd) [inline addon/components/impagination-collection.js]: Why the change in default load-horizon?

File: addon/components/impagination-collection.js [hunk 3]
```diff
@@ -28,6 +28,7 @@ export default Ember.Component.extend({
           Ember.run.debounce(this, 'flushQueue', offset, 1, false);
         }
       });
+      Object.defineProperty(collection, "length", { get: collection.length });
```
Comment (cowboyd) [inline addon/components/impagination-collection.js]: does just passing in `length: datasetState.length` into `CollectionInterface.create` no make the nut?

File: addon/components/impagination-collection.js [hunk 4]
```diff
@@ -104,7 +105,14 @@ var CollectionInterface = Ember.Object.extend(Ember.Array, {
...
 
-  length: Ember.computed(function () {
+  slice(start, end) {
+    if(start < 0) { return []; }
+    let records = this.datasetState.records.slice(start, end);
+    this.objectReadAt(start);
+    return records || undefined;
+  },
+
+  length: function() {
```
Comment (cowboyd) [inline addon/components/impagination-collection.js]: This should be a simple delegation to `datasetState.length`

File: tests/dummy/app/styles/app.css [hunk 5]
```diff
@@ -187,3 +187,13 @@ body {
   background-color: black;
   opacity: .4;
 }
+
+.virtual-each {
```
Comment (cowboyd) [inline tests/dummy/app/styles/app.css]: Is there any reason we don't just add this as a pull request to `virtual-each` itself? If every app that uses `virtual-each` needs this CSS to function, then it feels like this should be gotten for free, merely by including `virtual-each` in your project
